i.MX RT10XX: Add PWM support

loicpoulain opened the pull request:
Add PWM driver for RT10xx MCUs

Commit:
dts: arm: nxp: rt: Add PWM nodes

i.MX RT10XX processors have four eFlexPWM modules, each containing
four 2-channels PWM submodules.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
drivers: pwm: Add NXP MCUX PWM driver

This driver handles eFlexPWM submodules via MCUX hal.
Each submodule has two channels (A/B).

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
ext: mcux: Add HAS_MCUX_PWM config

For SoCs having PWM support.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

Commit:
soc: nxp_imx: Add PWM config

Enable MCUX PWM driver if PWM config is selected and hardware
support it.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

MaureenHelm commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -304,6 +304,157 @@
+
+		flexpwm3: flexpwm@403e4000 {
+			compatible = "nxp,flexpwm";
+			reg = <0x403e4000 0x4000>;
```
Missing interrupts

loicpoulain replied:
fixed

MaureenHelm commented on drivers/pwm/Kconfig.mcux:
```diff
@@ -0,0 +1,125 @@
+
+config PWM_MCUX_0
+	bool "NXP MCUX PWM 0 Output"
+	  depends on PWM_MCUX
```
Wrap all the instance configs with a big `if PWM_MCUX...endif` instead of repeating this `depends on`.

loicpoulain replied:
done

MaureenHelm commented on drivers/pwm/pwm_mcux.c:
```diff
@@ -0,0 +1,226 @@
+
+	duty_cycle = 100 * pulse_cycles / period_cycles;
+
+	/* FIXME: Force re-setup even for duty-cycle update */
```
Why?

loicpoulain replied:
fixed

loicpoulain replied:
could you please review again

MaureenHelm reviewed (changes requested):
Please enable pwm samples and tests too.

Commit:
boards: mimxrt1064: Add PWM support

PWM on GPIO_AD_B0_09 (USER_LED/Arduino J22 pin 5)

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

MaureenHelm commented on soc/arm/nxp_imx/rt/dts_fixup.h:
```diff
@@ -127,4 +127,100 @@
 
+#define DT_PWM_MCUX_0_BASE_ADDRESS		DT_NXP_FLEXPWM_403DC000_BASE_ADDRESS
+#define DT_PWM_MCUX_0_IRQ			DT_NXP_IMX_PWM_403DC000_PWM0_IRQ_0
+#define DT_PWM_MCUX_0_IRQ_PRI			DT_NXP_IMX_PWM_403DC000_PWM0_IRQ_0_PRIOTIRY
```
PRIOTIRY is misspelled

loicpoulain replied:
fixed

MaureenHelm reviewed (changes requested):
> Please enable pwm samples and tests too.

Would still like to see this

Commit:
boards: mimzzrt1064_evk: Add pwm-led0 alias

Used to build/run blink_led sample.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
> > Please enable pwm samples and tests too.
> 
> Would still like to see this

Done, at least form rt1064 for now

loicpoulain commented:
rebased

loicpoulain commented:
@MaureenHelm , could you please review?

MaureenHelm commented:
> > > Please enable pwm samples and tests too.
> > 
> > 
> > Would still like to see this
> 
> Done, at least form rt1064 for now

It doesn't work:
```
***** Booting Zephyr OS build v1.14.0-rc3-2405-ga061dbda9cc8 *****
PWM demo app-blink LED
pwm pin set fails
[00:00:00.008,000] <err> pwm_mcux: Too long period (1171875), adjust pwm prescaler!
```
Will you please fix?

loicpoulain commented:
> Will you please fix?

@MaureenHelm , well PWM source clock is 150MHz and maximum pwm clock prescaler is 128, which give a minimal clock of 1171875 Hz for the PWM. With the 16-bit resolution we can then have a maximum period of 0.05 second (2^16/1171875 = 0.05) whereas sample requests up to 1 second period. I can propbably modify the sample or at least customize for mimx board. what do you think?

Commit:
sample: blink_led: Adjust PWM period

The maximum period a PWM controller can generate depends on its input
clock and its resolution (16-bit, 32-bit...). Setting a 1s max period
generates error with some hardwares (mimxrt1064_evk) because of clock
limitation.

This patch fixes this issue by calculating maximum period of the PWM
device to adjust max and min periods accordingly.

Signed-off-by: Loic Poulain <loic.poulain@linaro.org>

loicpoulain commented:
@MaureenHelm , fixed now

loicpoulain commented:
hi @MaureenHelm , could you please review again?
